Initial Project Setup & Implementation

it-dinara opened the pull request:
## 📌 Summary
This PR introduces the **Adverly** project, built entirely from scratch. It includes both frontend and backend services, structured using **React, Redux, Docker**, and **Zod validation**. The goal is to establish a scalable foundation and gather feedback on architecture and implementation choices.

## 🏗️ Project Overview
- **Frontend:** React with Redux for state management.
- **Backend:** [Specify tech stack, e.g., Node.js, Express]
- **Docker:** Multi-service setup with health checks.
- **Validation:** Using `react-hook-form` + Zod.

## 🔍 Key Features Implemented
- ✅ **State management:** Redux integration with async queries.
- ✅ **Form handling:** Custom hooks (`useReduxFormSync`) for validation.
- ✅ **Dockerization:** Containers for frontend/backend with auto-build.
- ✅ **Git workflows:** Structured branching for future development.
- ✅ **Styling:** Basic UI setup for forms and layout.
- ✅ **Documentation:** Initial `README.md` for setup instructions.

## 📖 Testing Instructions
1. Clone the repo:
   ```sh
   git clone https://github.com/it-dinara/adverly.git
   cd adverly
   
   
## ❓ Questions for Review
- Are the Redux patterns efficient and scalable?
- Does the project structure align with best practices?
- Any suggestions for improving form handling or API interactions?
- Feedback on Docker setup and environment variables?

it-dinara committed:
start

it-dinara committed:
car inputs

it-dinara committed:
auto categories work

it-dinara committed:
-a

it-dinara committed:
react-hook-form and zod

it-dinara committed:
car model field works

it-dinara committed:
Steps without Redux, only Sessionstorage

it-dinara committed:
new categories added

it-dinara committed:
debounce

it-dinara committed:
back to redux

it-dinara committed:
Auto, RealEstate, and FirstStep components were updated to use the new SecondStep component. The SecondStepForm module was renamed to SecondStep, and the corresponding CSS file was also renamed. The FormPage and ItemPage components were updated to reflect these changes. The formSlice and form types were also updated accordingly.

it-dinara committed:
Refactor Auto and Services components; Update resetForm

it-dinara committed:
Fix type coercion for area, rooms, and price in RealEstate component

it-dinara committed:
Refactor ItemPage component to improve structure and readability of item details rendering

it-dinara committed:
Add custom useReduxFormSync hook for form state synchronization with Redux

it-dinara committed:
Refactor form handling across Auto, RealEstate, and Services components; implement useReduxFormSync hook for improved state management and validation. Add default form values and validation schemas. Update ItemPage and formSlice to accommodate new structure. Enhance TypeScript configurations for better module resolution.

it-dinara committed:
Refactor useReduxFormSync and useSubmitForm hooks for improved state management; implement debouncing in useReduxFormSync. Remove unnecessary resetForm dispatch in ItemsListPage. Enhance debounce utility for better functionality.

it-dinara committed:
Add 'mode: onBlur' to useReduxFormSync in Auto, RealEstate, and Services components for improved form validation handling

it-dinara committed:
Refactor form handling and validation across Auto, RealEstate, and Services components; update schemas and default values for improved structure and consistency. Enhance useReduxFormSync hook for better state management and form synchronization.

it-dinara committed:
Refactor Auto component and form handling; update form state structure to use selectedCategoryForm for brand, model, year, and mileage. Modify useSubmitForm to accommodate new data structure and adjust ItemPage to reflect category changes. But I didn't complete this whole refactoring, The whole approache doesn't make sence compare to one I have already done, so, I'm comming back to prev one.

it-dinara committed:
Refactor form handling across Auto, RealEstate, and Services components; update form state structure to simplify data management. Enhance form validation and error handling, and improve default values for better consistency. Adjust hooks for improved state management and submission processes. STRUCTURED

it-dinara committed:
Refactor form components and styles; remove default margins in CSS, update category handling in form steps, and enhance validation schemas for improved user experience. Implement memoization for performance optimization and ensure consistent default values across forms.

it-dinara committed:
Refactor RealEstate styles; remove margin from form for better alignment. Update formDefaults to clean up unused properties. Enhance ItemsListPage filter options structure for improved clarity and type safety. Remove redundant types from form.ts for cleaner code.

it-dinara committed:
Refactor axios instance imports; update references from "AxiosInstance" to "AxiosAdverly" across multiple files. Introduce item types and state interfaces in itemList.ts for improved type safety and structure.

it-dinara committed:
Remove unused CategoryButton and Input components, along with related CSS and constants, to streamline the codebase. Add initial README.md for project documentation and setup instructions.

it-dinara committed:
Remove item deletion functionality and associated UI elements from FormPage

it-dinara committed:
Reset branch

it-dinara committed:
Merge branch 'main' into feature-branch

it-dinara committed:
Update .gitignore to ignore entire frontend/dist/ directory instead of specific bundle.js file

it-dinara committed:
Stop tracking dist and frontend/dist/

Andrushka36 commented on README.md:
```diff
@@ -0,0 +1,17 @@
+Ensure you have the following installed:
+- [Docker](https://www.docker.com/get-started)
+- [Docker Compose](https://docs.docker.com/compose/install/)
+- [Node.js](https://nodejs.org/)
```
Вижу тут какое-то противоречие. Если мы запускаем наше приложение внутри докера, то зачем нам локально нода?

it-dinara replied:
Можно и так и так запускать

Andrushka36 commented on README.md:
```diff
@@ -0,0 +1,17 @@
+## 🏗️ Setup & Build
+To set up and build the project:
+
+1. Clone the repository:
```
~Я бы тут, на самом деле, еще команды для запуска приложения добавил. Плюс, наверное, даже как для запуска приложения внутри докера, так и просто типа `npm i` / `npm start`.~

Andrushka36 commented on frontend/src/axios-adverly.ts:
```diff
@@ -0,0 +1,7 @@
+import axios from "axios";
+
+const instance = axios.create({
+  baseURL: "http://127.0.0.1:3000/",
```
А тут обязательно написать хост? Кажется, такое не очень удобно, если ты будешь деплоить куда-то приложение.

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.module.css:
```diff
@@ -0,0 +1,72 @@
+  cursor: pointer;
+}
+
+/* --------- ERROR STYLE --------- */
```
Кажется, этот комментарий избыточный, потому что название класса ниже однозначно говорит, про что он. В целом, на самом деле, я считаю, что комментариев в коде должно быть мало и только там, где это обусловлено какой-то бизнес-логикой приложения либо наличием какого-то хака.

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.tsx:
```diff
@@ -0,0 +1,191 @@
+  const { isEditing } = useAppSelector((state) => state.form);
+  const [carBrands, setCarBrands] = useState<Car[]>([]);
+
+  // Fetch car brands data once on mount
```
Вот тут тоже комментарий не нужен, потому что у тебя в эффекте вызывается функция `fetchCarBrands`, и по названию понятно, что она делает.

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.tsx:
```diff
@@ -0,0 +1,191 @@
+  const selectedModel = watch("model") || "";
+
+  // Memoize filtered models so we don’t re-compute on every render
+  const models = useMemo(
```
`useMemo` лучше использовать, чтобы избежать ненужных ререндеров или каких-то тяжелых вычислений.

it-dinara committed:
Add HtmlWebpackPlugin and update index.html structure

it-dinara committed:
Add HtmlWebpackPlugin and update index.html structure

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.module.css:
```diff
@@ -0,0 +1,72 @@
+
+.button {
+  padding: 0.7rem 1.5rem;
+  background-color: #0077cc;
```
Вижу, что этот цвет используется в этом файле не только в этом селекторе, поэтому, возможно, было бы неплохо положить его в какую-нибудь CSS-переменную. Плюс если ты не используешь сторонние библиотеки компонентов, то свой набор CSS-переменных (для всего приложения в целом) реально круто смотрелся бы.

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.tsx:
```diff
@@ -0,0 +1,191 @@
+          ""
+        }
+      >
+        {selectedModel &&
```
Если честно, мне кажется, избыточным добавление активного элемента наверх. Лучше для `option` задать атрибут `selected`, хотя, возможно, `register `для `select` уже как-то добавляет `selected`.

it-dinara replied:
он нужен чтобы сбрасывалось поле модели авто если выбран другой бренд

Andrushka36 replied:
Хм. Что-то я не совсем понимаю, как это работает. Если у тебя в `selectedModel` ничего нет, то отрисуется `<option value="">Выберите модель</option>`, а если есть, то  `<option value="__name__">__name__</option>` - так ведь?
Кстати, в `value`, наверное, все-таки лучше держать `id`, а не `name`, потому что если бы у тебя был настоящий бекенд с базой данных, то там в объявлении хотелось бы именно идентификатор сохранять.

it-dinara replied:
"Что-то я не совсем понимаю, как это работает. Если у тебя в selectedModel ничего нет, то отрисуется <option value="">Выберите модель</option>, а если есть, то <option value="__name__">__name__</option> - так ведь?"

Да,но у каждого бренда свой список моделей и юзер может выбрать бренд и список моделей обновляется, затем выбирает модель. А потом юзер может опять поменять бренд, но модель останется от другого бренда. То есть юзер выбирает бренд BMW и затем модель 500,а затем вернется и поменяет бренд на Mercedes, а модель останется от BMW, то есть 500, но её не существует у Mercedes. Так что я проверяю есть ли эта модель у Mercedes и если нет, то сбрасываю значение до дефолтного.

Andrushka36 replied:
Ага, понял. Но вообще при изменении бренда тебе всегда нужно обнулять марку, потому что ведь при текущей реализации ты как бы из интерфейса убираешь подпись, но из стейта не удаляешь?

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.tsx:
```diff
@@ -0,0 +1,191 @@
+          </div>
+
+          <div className={s.formGroup}>
+            <label className={`${s.label} ${s.required}`} htmlFor="model">
```
На самом деле, для выпадающего списка прям так и хочется какой-то общий компонент создать, чтоб не копипастить.

Andrushka36 commented on frontend/src/components/Categories/RealEstate/RealEstate.tsx:
```diff
@@ -0,0 +1,98 @@
+  [key: string]: any; // for other props like register
+};
+
+const RealEstate: React.FC<RealEstateProps> = ({ register, errors }) => {
```
Этот компонент очень похож на `Auto `и на `Services`. Тут тоже как-нибудь хотелось бы избежать дублирование кода.

Andrushka36 commented on frontend/src/components/Categories/RealEstate/RealEstate.tsx:
```diff
@@ -0,0 +1,98 @@
+import { useAppSelector } from "Redux/hooks";
+
+const propertyTypes = [
+  "Квартира",
```
Автошные марки и модели ты запрашиваешь с бекенда - это прикольно. Здесь тоже можно было бы какую-нибудь ручку сделать.

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.tsx:
```diff
@@ -0,0 +1,191 @@
+  useEffect(() => {
+    const fetchCarBrands = async () => {
+      try {
+        const { data } = await axios.get<Car[]>("/cars.json");
```
Я бы, кстати, предложил прям именно эндпоинт на бекенде создать по аналогии с теми, что уже есть.

Andrushka36 commented on frontend/src/components/Categories/Auto/Auto.tsx:
```diff
@@ -0,0 +1,191 @@
+  [key: string]: any; // for any other props
+};
+
+const Auto: React.FC<AutoProps> = ({ register, watch, errors }) => {
```
Ты здесь используешь 3 параметра, при этом в `AutoProps `у тебя определены 4 плюс `[key: string]: any` - почему так? Плюс от `any `лучше отказаться вообще.

Andrushka36 commented on frontend/src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,36 @@
+const Pagination: React.FC = () => {
+  const dispatch = useAppDispatch();
+  const { currentPage, items, itemsPerPage } = useAppSelector(
+    (state) => state.items
```
Селектор лучше вынести за пределы компоненты.

Andrushka36 commented on frontend/src/components/Steps/FirstStep/FirstStep.tsx:
```diff
@@ -0,0 +1,124 @@
+        {isEditing ? "Редактирование объявления" : "Создание объявления"}
+      </h2>
+      <p className={s.description}>Заполните форму</p>
+      <h2 className={s.subtitle}>Основной шаг</h2>
```
2 `h2 `на одном уровне, при этом 0 `h1 `- такое себе. По семантике первый `h2 `должен быть `h1`, потому что характеризует всю страницу.

it-dinara committed:
Update README.md to enhance setup instructions and clarify development process

it-dinara committed:
refactor: clean up package.json by removing unused dependencies

Andrushka36 commented on frontend/src/components/Steps/SecondStep/SecondStep.tsx:
```diff
@@ -0,0 +1,30 @@
+  } else if (selectedCategory === Categories.SERVICES) {
+    categoryForm = <Services {...props} />;
+  }
+  return <>{categoryForm}</>;
```
Можно поспорить, но мне кажется, что запись
```
const selectedCategory = useAppSelector((state) => state.form.category);

if (selectedCategory === Categories.AUTO) {
  return <Auto {...props} />;
}

if (selectedCategory === Categories.REAL_ESTATE) {
  return <RealEstate {...props} />;
}

if (selectedCategory === Categories.SERVICES) {
  return <Services {...props} />;
}

return null;
```

выглядит лучше, потому что ты сразу видишь, что в зависимости от значения `selectedCategory` рендерится разный компонент. Лучше стараться делать `return `в функции (неважно, реактовский ли компонент-функция это или обычная функция) как можно быстрее, потому что при дебаге в случае единственного `return `в конце тебе нужно внимательно следить за тем, что происходит в твоей функции, особенно если у тебя много разных ифаков.

Andrushka36 commented on frontend/src/pages/FormPage.tsx:
```diff
@@ -0,0 +1,72 @@
+  const fetchItems = async () => {
+    try {
+      const response = await axios.get<{ id: number; name: string }[]>(
+        `http://127.0.0.1:3000/items`
```
Тут бы тоже без хоста обойтись.

Andrushka36 commented on frontend/src/pages/FormPage.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  const fetchItems = async () => {
+    try {
+      const response = await axios.get<{ id: number; name: string }[]>(
```
Аналогичный тип у тебя в `useState` - было бы неплохо вынести его, чтоб не копипастить. Плюс я не очень люблю таким образом инлайнить тип, потому что он как-то немного отвлекает что ли от самой логики.

Andrushka36 commented on frontend/src/pages/ItemPage.tsx:
```diff
@@ -0,0 +1,98 @@
+  };
+
+  // Create a list of item details
+  let list = Object.entries(item ?? {}).map(([type, value]) => {
```
А почему тут переменная создана через `let`? Мы ее потом изменяем как-то? И вообще, а можно тут без переменной обойтись?

Andrushka36 commented on frontend/src/pages/ItemPage.tsx:
```diff
@@ -0,0 +1,98 @@
+    );
+  });
+
+  let photo = item && (
```
Аналогичный вопрос тут.

Andrushka36 commented on frontend/src/pages/ItemPage.tsx:
```diff
@@ -0,0 +1,98 @@
+    </div>
+  );
+
+  let itemDetails = (
```
Тут, мне кажется, тоже можно без переменной обойтись.

Andrushka36 commented on frontend/src/pages/ItemPage.tsx:
```diff
@@ -0,0 +1,98 @@
+  return (
+    <div className={s.container}>
+      {error && <p className={s.errorMessage}>{error}</p>}
+      {item && !error ? itemDetails : null}
```
`item && !error && itemDetails`?

Andrushka36 commented on frontend/src/pages/ItemsListPage.tsx:
```diff
@@ -0,0 +1,225 @@
+      </div>
+
+      <Link to={`/form`} className={s.submitButtonLink}>
+        <button type="button" className={s.submitButton}>
```
Вкладывать кнопку в ссылку не стоит. По семантике тут у тебя должна быть ссылка, потому что осуществляется переход на другую страницу.

Andrushka36 commented on frontend/src/redux/slices/itemSlice.ts:
```diff
@@ -0,0 +1,67 @@
+export const fetchItem = createAsyncThunk<ItemDetails, string>(
+  "Item/fetchItem",
+  async (id: string): Promise<ItemDetails> => {
+    console.log("suka bleat");
```
Перед пушем `console.log`, особенно такие, лучше удалять, конечно. :)

Andrushka36 commented on frontend/src/redux/slices/itemSlice.ts:
```diff
@@ -0,0 +1,67 @@
+  "Item/fetchItem",
+  async (id: string): Promise<ItemDetails> => {
+    console.log("suka bleat");
+    const response = await axios.get<ItemDetails>(
```
А тут нельзя сделать `axiosInstance.get`?

Andrushka36 commented on frontend/src/redux/slices/itemSlice.ts:
```diff
@@ -0,0 +1,67 @@
+  }
+);
+
+const ItemSlice = createSlice({
```
`itemSlice`.

Andrushka36 commented on frontend/src/redux/slices/itemsListSlice.ts:
```diff
@@ -0,0 +1,36 @@
+
+// Async thunk to fetch items (simulate API call)
+export const fetchItems = createAsyncThunk("items/fetchItems", async () => {
+  const response = await axios.get<Item[]>("http://127.0.0.1:3000/items");
```
Тут тоже поинт про хост.

it-dinara committed:
refactor: redirect from base root to form.
Updated the React Router configuration to automatically redirect the root URL ("/")
to "/form", improving consistency in navigation.

Andrushka36 commented on npm:
```diff
@@ -0,0 +1,8 @@
+      PID    PPID    PGID     WINPID   TTY         UID    STIME COMMAND
```
Что это за файл?

it-dinara committed:
refactor: update routing paths and remove unused item slice

it-dinara committed:
Refactor form components: move RealEstate, Auto, Services components, consolidate styles.
Move FirstStep and SecondStep components for improved structure. Clean up CSS files.

it-dinara committed:
refactor: move FirstStep and SecondStep components, remove unused query in FormPage  component, updated props structures in SecondStep component. Updated tsconfig include field to avoid redundancy.

it-dinara committed:
refactor: update form components to use specific types for props, improve error handling.

it-dinara committed:
refactor: add dotenv for environment variable management, update axios instance to use REACT_APP_API_URL

it-dinara committed:
refactor: remove redundant useMemo hook

it-dinara committed:
refactor: extract model selection logic into a function  in Auto component

it-dinara committed:
refactor: extract pagination state selection into a dedicated function

it-dinara committed:
refactor: change variable declarations from let to const Item component

it-dinara committed:
refactor: simplify item details rendering logic in ItemPage component

it-dinara committed:
refactor: remove submit button styles and update link to display text directly

it-dinara committed:
refactor: add npm to .gitignore to prevent tracking of npm files

it-dinara committed:
refactor: remove dotenv dependency from root package.json and package-lock.json
